### PR TITLE
mon/OSDMonitor: encode full osdmaps with features all OSDs can understand

### PR DESCRIPTION
**Pull request by liewegas**
Also prompt users to set the require_*_osds flags once tey finish their upgrades.

### CODE REVIEW

**Comment by jdurgin** [comment]
We should discuss automatically setting the flags - it's silly to add extra steps for users to jump through for every upgrade. Perhaps some heuristic like 'after all up osds are kraken for 24h, require the kraken feature bit'.

**Comment by liewegas** [comment]
Yeah.  It's a bit tricky because we need to avoid situations where all the un-upgraded osds crash or are stopped and only upgraded osds are up.  It could be gated on the cluster being healthy, perhaps... or all in osds being up.

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -1220,13 +1220,25 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
     }
   }
 
+  // determine appropriate features
+  uint64_t features = mon->quorum_features;
+  if (!newmap.has_flag(CEPH_ODSMAP_REQUIRE_JEWEL)) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: s/ODS/OSD/

**Comment by tchaikov** [comment]
lgtm.

@liewegas we need run this thru at least one upgrade test, and probably you could include https://github.com/tchaikov/ceph/commit/31233d6b2a4666454a24f7dc2ea9cf6eb9969741 in your PR? and try to remove 'failed to encode map' from (at least some of) the upgrade tests. i will take care of this once this change is merged.

**Comment by tchaikov** [comment]
do we want to backport this change to jewel?

**Comment by liewegas** [comment]
Yeah, I think we should backport.

**Comment by tchaikov** [comment]
@liewegas how's test going?

**Comment by liewegas** [comment]
trying to get it through upgrade/jewel-x

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2016-10-15_14:11:58-upgrade:jewel-x-wip-sage-testing---basic-smithi/

**Pull request merged**